pageserver/src/tenant/remote_timeline_client/index.rs

```diff
@@ -184,8 +193,67 @@ impl From<&LayerFileMetadata> for IndexLayerMetadata {
     }
 }
 
+/// Limited history of earlier ancestors.
+///
+/// A timeline can have more than 1 earlier ancestor, in the rare case that it was repeatedly
+/// reparented by having an later timeline be detached from it's ancestor.
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize, Default)]
+pub(crate) struct Lineage {
+    /// Has the `reparenting_history` been truncated to `Lineage::REMEMBER_AT_MOST`.
+    #[serde(skip_serializing_if = "is_false", default)]
+    reparenting_history_truncated: bool,
+    /// Earlier ancestors, truncated when `reparented_overflown`
+    #[serde(skip_serializing_if = "Vec::is_empty", default)]
+    reparenting_history: Vec<TimelineId>,
+    /// The ancestor from which this timeline has been detached from and when.
+    #[serde(skip_serializing_if = "Option::is_none", default)]
+    original_ancestor: Option<(TimelineId, Lsn, NaiveDateTime)>,
```

Comment:
Doesn't using an `Option` here limit us in a potential future where we allow detaching timelines from a not-root ancestor? In that future, we'd want to record a series of detachments, there isn't just _one_ original.

Reply:
I think that detaching would be just a detaching from a single ancestor, but having all of the layers merged from the ancestry (with ancestor scoped lsn visibility)...

But perhaps it is limiting. I don't see a problem for adding multiple in the future, and I don't think we are going to add the N-step either.

Reply:
Fixed with [f2853dc](https://github.com/neondatabase/neon/pull/7639/commits/f2853dca91ad931d9b68ebf0674ddcd2623cefcf)